### ISSUE TITLE
Segfault when hitting ^C on backtrace

### ISSUE THREAD

**Issue by jakebolewski** [JuliaLang/julia#17744]
Steps to reproduce:

checkout the latest master version of JuliaParser
run the testshell in bin/testshell.jl
enter `"""` on the testshell parse line and hit ^C immediately afterwards

```julia
v0.5/JuliaParser [master●] » julia-dev bin/testshell.jl
INFO: Recompiling stale cache file /Users/jacobbolewski/.julia/lib/v0.5/AbstractTrees.ji for module AbstractTrees.
INFO: Recompiling stale cache file /Users/jacobbolewski/.julia/lib/v0.5/JuliaParser.ji for module JuliaParser.
INFO: Recompiling stale cache file /Users/jacobbolewski/.julia/lib/v0.5/Compat.ji for module Compat.
INFO: Recompiling stale cache file /Users/jacobbolewski/.julia/lib/v0.5/TerminalUI.ji for module TerminalUI.
INFO: Recompiling stale cache file /Users/jacobbolewski/.julia/lib/v0.5/Reactive.ji for module Reactive.
INFO: Recompiling stale cache file /Users/jacobbolewski/.julia/lib/v0.5/VT100.ji for module VT100.
INFO: Recompiling stale cache file /Users/jacobbolewski/.julia/lib/v0.5/Colors.ji for module Colors.
INFO: Recompiling stale cache file /Users/jacobbolewski/.julia/lib/v0.5/FixedPointNumbers.ji for module FixedPointNumbers.
INFO: Recompiling stale cache file /Users/jacobbolewski/.julia/lib/v0.5/ColorTypes.ji for module ColorTypes.
INFO: Recompiling stale cache file /Users/jacobbolewski/.julia/lib/v0.5/Reexport.ji for module Reexport.
parse > """
^C
signal (2): Interrupt: 2
while loading /Users/jacobbolewski/.julia/v0.5/JuliaParser/bin/testshell.jl, in expression starting on line 115
_ZNK4llvm6Module13getModuleFlagENS_9StringRefE at /Users/jacobbolewski/Julia/julia-dev/usr/lib//libLLVM-3.7.dylib (unknown line)
jl_setup_module at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:896
emit_function at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:4018
jl_compile_linfo at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:809
emit_invoke at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:2688 [inlined]
emit_expr at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3166
emit_assignment at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:2927 [inlined]
emit_expr at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3189
emit_stmtpos at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3068
emit_function at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:4731
jl_compile_linfo at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:809
emit_invoke at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:2688 [inlined]
emit_expr at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3166
emit_call at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:2754
emit_expr at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3176
emit_call at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:2754
emit_expr at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3176
emit_assignment at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:2927 [inlined]
emit_expr at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3189
emit_stmtpos at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3068
emit_function at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:4731
jl_compile_linfo at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:809
emit_invoke at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:2688 [inlined]
emit_expr at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3166
emit_function at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:4695
jl_compile_linfo at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:809
emit_invoke at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:2688 [inlined]
emit_expr at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3166
emit_assignment at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:2927 [inlined]
emit_expr at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3189
emit_stmtpos at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3068
emit_function at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:4731
jl_compile_linfo at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:809
emit_invoke at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:2688 [inlined]
emit_expr at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3166
emit_assignment at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:2981 [inlined]
emit_expr at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3189
emit_stmtpos at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:3068
emit_function at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:4731
jl_compile_linfo at /Users/jacobbolewski/Julia/julia-dev/src/codegen.cpp:809
jl_compile_for_dispatch at /Users/jacobbolewski/Julia/julia-dev/src/gf.c:1311
jl_call_method_internal at /Users/jacobbolewski/Julia/julia-dev/src/./julia_internal.h:173 [inlined]
jl_apply_generic at /Users/jacobbolewski/Julia/julia-dev/src/gf.c:1930
#parse at ./<missing>:0
parse at /Users/jacobbolewski/.julia/v0.5/JuliaParser/src/parser.jl:2154
jl_call_method_internal at /Users/jacobbolewski/Julia/julia-dev/src/./julia_internal.h:178 [inlined]
jl_apply_generic at /Users/jacobbolewski/Julia/julia-dev/src/gf.c:1930
do_call at /Users/jacobbolewski/Julia/julia-dev/src/interpreter.c:66
eval at /Users/jacobbolewski/Julia/julia-dev/src/interpreter.c:190
jl_toplevel_eval_flex at /Users/jacobbolewski/Julia/julia-dev/src/toplevel.c:558
jl_toplevel_eval_in_warn at /Users/jacobbolewski/Julia/julia-dev/src/builtins.c:576
eval at ./boot.jl:234
jl_call_method_internal at /Users/jacobbolewski/Julia/julia-dev/src/./julia_internal.h:178 [inlined]
jl_apply_generic at /Users/jacobbolewski/Julia/julia-dev/src/gf.c:1930
#7 at /Users/jacobbolewski/.julia/v0.5/JuliaParser/bin/testshell.jl:86
jl_call_method_internal at /Users/jacobbolewski/Julia/julia-dev/src/./julia_internal.h:178 [inlined]
jl_apply_generic at /Users/jacobbolewski/Julia/julia-dev/src/gf.c:1930
run_interface at ./LineEdit.jl:1570
RunShell at /Users/jacobbolewski/.julia/v0.5/JuliaParser/bin/testshell.jl:113
jl_call_method_internal at /Users/jacobbolewski/Julia/julia-dev/src/./julia_internal.h:178 [inlined]
jl_apply_generic at /Users/jacobbolewski/Julia/julia-dev/src/gf.c:1930
do_call at /Users/jacobbolewski/Julia/julia-dev/src/interpreter.c:66
eval at /Users/jacobbolewski/Julia/julia-dev/src/interpreter.c:190
jl_toplevel_eval_flex at /Users/jacobbolewski/Julia/julia-dev/src/toplevel.c:558
jl_parse_eval_all at /Users/jacobbolewski/Julia/julia-dev/src/ast.c:717
jl_load at /Users/jacobbolewski/Julia/julia-dev/src/toplevel.c:596 [inlined]
jl_load_ at /Users/jacobbolewski/Julia/julia-dev/src/toplevel.c:605
include_from_node1 at ./loading.jl:426
jl_call_method_internal at /Users/jacobbolewski/Julia/julia-dev/src/./julia_internal.h:178 [inlined]
jl_apply_generic at /Users/jacobbolewski/Julia/julia-dev/src/gf.c:1930
process_options at ./client.jl:262
_start at ./client.jl:318
jl_call_method_internal at /Users/jacobbolewski/Julia/julia-dev/src/./julia_internal.h:178 [inlined]
jl_apply_generic at /Users/jacobbolewski/Julia/julia-dev/src/gf.c:1930
true_main at /Users/jacobbolewski/Julia/julia-dev/usr/bin/julia (unknown line)
main at /Users/jacobbolewski/Julia/julia-dev/usr/bin/julia (unknown line)
unknown function (ip: 0xffffffffffffffff)
Allocations: 2503100 (Pool: 2501372; Big: 1728); GC: 2
[1]    4323 segmentation fault  /Users/jacobbolewski/Julia/julia-dev/usr/bin/julia bin/testshell.jl
```

**Comment by yuyichao** [comment]
Dup of https://github.com/JuliaLang/julia/issues/17706

**Comment by jakebolewski** [comment]
No, I still get a segfault even with the latest master.  I've updated the original backtrace in the issue now that I have rebuilt julia.

```
julia> versioninfo()
Julia Version 0.5.0-rc0+113
Commit fc3cbc9 (2016-08-01 20:53 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin15.6.0)
  CPU: Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, haswell)
```

**Comment by yuyichao** [comment]
> No, I still get a segfault even with the latest master

So? #17706 is still open.

**Comment by jakebolewski** [comment]
@yuyichao from my reading of that issue, does this mean that all custom REPL modes have to add the appropriate `sigatomic_begin()`, `sigatomic_end()` statements?  If that is the case feel free to close.

**Comment by yuyichao** [comment]
There's a few issues here and the segfault cannot be reliably fixed even when #17706 is closed.

1. Any code that don't want to be interrupted should be wrapped in `disable_sigint` and use `reenable_sigint` on the part that can handle `InterruptException`. Note that `sigatomic_begin` and `sigatomic_end` are internal API and their semantics can change at any time. This usually includes the error handling part of REPL.
2. Any code that want to handle `InterruptException` instead of existing asynchronously should ccall `jl_exit_on_sigint(0)`. This is done in the normal REPL mode.
3. #17727 makes the exit less-segfaulty but there's no way this can be guaranteed. In this particular case, it spends a lot of time in LLVM and the exit is almost certainly going to segfault. One option is to always use `_exit` to exit on SIGINT, I'm not sure which one is better (segfault sometimes but does cleanup otherwise, or never segfault and never does cleanup).

**Comment by jakebolewski** [comment]
Thanks, those bullet points are a helpful synthesis of that thread.  Maybe you could copy and paste those to the devdocs so they don't get lost in a morass of issue comments?

Sorry for the noise, I have not been keeping up with issues at all.
